ops/testing.py

```diff
@@ -1759,6 +1789,104 @@ class NonAbsolutePathError(Exception):
     """
 
 
+class _MockStorage:
+    def __init__(self, location, src):
+        self._src = src
+        os.makedirs(src, exist_ok=True)
+        self._location = location
+
+    def contains(self, path):
+        return str(path).startswith(self._location)
+
+    def _srcpath(self, path):
+        suffix = path[len(self._location):]
+        if suffix[0] == '/':
+            suffix = suffix[1:]
+        srcpath = os.path.join(self._src, suffix)
+        return srcpath
+
+    def create_dir(self, path: str, make_parents: bool = False, **kwargs) -> '_Directory':
+        assert(self.contains(path))
+        srcpath = self._srcpath(path)
+
+        dirname = os.path.dirname(srcpath)
+        if not os.path.exists(dirname):
+            if not make_parents:
+                raise RuntimeError('no such directory {}'.format(dirname))
+            os.makedirs(dirname)
+        os.makedirs(srcpath)
+        return _Directory(self, path, **kwargs)
+
+    def create_file(
+            self,
+            path: str,
+            data: typing.Union[bytes, str, typing.BinaryIO, typing.TextIO],
+            encoding: typing.Optional[str] = 'utf-8',
```

Comment:
I'd expect anything marked as optional to have None as default,
I'd do:

- `encoding: str = None`
- specify in the docstring that the default is 'utf-8'
- later do: `encoding = encoding or 'utf-8'`

However it seems to me that you really do need an encoding one way or another, so why not just
`encoding: str = 'utf-8'`

Reply:
Yeah - this was a copy-paste holdover.  Fixed.